doc/man/gnunet-config.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2021 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd July 15, 2012
.Dt GNUNET-CONFIG 1
.Os
.Sh NAME
.Nm gnunet-config
.Nd manipulate GNUnet configuration files and provide information on how GNUnet was built
.Sh SYNOPSIS
.Nm
.Op Fl b Ar BACKEND | Fl -supported-backend= Ns Ar BACKEND
.Op Fl C | -cflags
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl F | -full
.Op Fl f | -filename
.Op Fl h | -help
.Op Fl j | -libs
.Op Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
.Op Fl o Ar OPTION | Fl -option= Ns Ar OPTION
.Op Fl p | -prefix
.Op Fl r | -rewrite
.Op Fl S |-list-sections
.Op Fl s Ar SECTION | Fl -section= Ns Ar SECTION
.Op Fl V Ar VALUE | Fl -value Ar VALUE
.Op Fl v | -version
.Sh DESCRIPTION
.Nm
can be used to read or modify GNUnet configuration files.
It can also provide informations to properly build applications on top of
GNUnet, like appropriate values for CFLAGS or the installation prefix.
.Bl -tag -width indent
.It Fl b Ar BACKEND | Fl -supported-backend= Ns Ar BACKEND
Tests whether the specified BACKEND is supported by the current installation.  The backend must match the name of a plugin, e.g. "namestore_postgres" for the Postgres database backend of the "NAMESTORE" service. If the BACKEND is supported, gnunet-config will return a status code of 0 (success), otherwise 77 (unsupported).  When this option is specified, no other options may be specified. Specifying this option together with other options will cause gnunet-config to return a status code of 1 (error).
.It Fl C | -cflags
Print an appropriate value for CFLAGS to compile applications on top of GNUnet.
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Use the configuration file FILENAME.
.It Fl F | -full
Write the full configuration, not just the deltas to the defaults.
.It Fl f | -filename
Try to perform expansions as if the option values represent filenames (will also be applied even if the option is not really a filename).
.It Fl h | -help
Print short help on options.
.It Fl j | -libs
Print an appropriate value for LIBS to compile applications on top of GNUnet.
.It Fl L Ar LOGLEVEL | Fl -loglevel= Ns Ar LOGLEVEL
Use LOGLEVEL for logging.
Valid values are DEBUG, INFO, WARNING and ERROR.
.It Fl o Ar OPTION | Fl -option= Ns Ar OPTION
Which configuration option should be accessed or edited.
Required to set a value.
If not given, all values of a given section will be printed in the format "OPTION = VALUE".
.It Fl p | -prefix
Print GNUnet's installation prefix (for example /usr/local.)
.It Fl r | -rewrite
Write the configuration to file, even if it did not change.
.It Fl S |-list-sections
List available configuration sections for use with --section.
.It Fl s Ar SECTION | Fl -section= Ns Ar SECTION
Which configuration section should be accessed or edited.
Required option.
.It Fl V Ar VALUE | Fl -value Ar VALUE
Configuration value to store in the given section under the given option.
Must only be given together with -s and -o options.
.It Fl v | -version
Print GNUnet version number.
.El
.Sh SEE ALSO
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


ChangeLog
Wed, 1 Sep 2021 10:28:09 +0200 (d51000320)
add fix for GNUnet directory permissions by madmurphy - Christian Grothoff

Mon, 30 Aug 2021 09:35:40 +0200 (9eb417e72)
exec' - Christian Grothoff

Sat, 28 Aug 2021 21:50:06 +0200 (2994fa434)
improve error message and inherit stderr as suggested by Florian - Christian Grothoff

Sat, 28 Aug 2021 18:42:53 +0200 (c6528ac77)
debian: systemd installation - Florian Dold

Sat, 28 Aug 2021 18:31:33 +0200 (eacbf5923)
debian: clean up - Florian Dold

Sat, 28 Aug 2021 18:27:45 +0200 (ec6b80f07)
debian: more fixes and formatting - Florian Dold

Sat, 28 Aug 2021 18:27:11 +0200 (830bc7ab2)
debian: more fixes - Florian Dold

Sat, 28 Aug 2021 18:23:05 +0200 (941ea7f20)
debian: various packaging fixes - Florian Dold

Sat, 28 Aug 2021 17:46:26 +0200 (b317b3e11)
use ISO 8609 / RFC 3339 timestamps for logging - Florian Dold

Sat, 28 Aug 2021 16:54:38 +0200 (e74fadf93)
release GNUnet 0.15.3 - Christian Grothoff

Sat, 28 Aug 2021 12:54:32 +0200 (1757980da)
partial fix for problems raised by Alessio on the ML - Christian Grothoff

Sat, 28 Aug 2021 12:50:20 +0200 (66e906c25)
partial fix for problems raised by Alessio on the ML - Christian Grothoff

Wed, 25 Aug 2021 12:52:12 +0200 (45532e7bc)
further simplify libgnunetpq event API and implementation, also trigger events on 'loopback' that are not socket-activated - Christian Grothoff

Tue, 24 Aug 2021 16:12:34 +0200 (a0e689f5e)
debian: libgnunet-dev should not depend on GNUnet - Florian Dold

Sun, 8 Aug 2021 22:10:55 +0200 (b71740271)
debian: do not pollute skel (see Debian Policy Section 10.7.5.) - Florian Dold

Sun, 8 Aug 2021 21:20:09 +0200 (2de19bb85)
debian: leave out debian version '0', as lintian complains - Florian Dold

Sun, 8 Aug 2021 21:14:42 +0200 (788f1edbe)
debian: bump version (0.15.0-1) - Florian Dold

Fri, 6 Aug 2021 11:56:36 +0200 (564a1945e)
debian: bump version - Florian Dold

Wed, 4 Aug 2021 20:41:25 +0200 (ff5075e7f)
do not generate tautological comparison warnings for assertions (if gcc/clang version supports them) - Christian Grothoff

Wed, 4 Aug 2021 15:09:17 +0200 (873381c65)
first cmd for a simple tng test running. part of cleanup and sanity check when starting missing - t3sserakt

Sun, 1 Aug 2021 16:23:32 +0200 (372516f77)
0 for absolute time is special, and should be treated as such in JSON generation - Christian Grothoff

Sun, 1 Aug 2021 14:53:14 +0200 (c98bf47a8)
PQ: fix notification API bugs - Christian Grothoff

Sun, 1 Aug 2021 12:26:20 +0200 (2032f3ce3)
debian: compat level, removal of user units, tweaks - Florian Dold

Sat, 31 Jul 2021 22:42:45 +0200 (d02ea47f5)
JSON: allow NULL-passing to GNUNET_JSON_pack_data_varsize - Christian Grothoff

Sat, 31 Jul 2021 20:04:51 +0200 (be8bbf724)
debian: bump version - Florian Dold

Sat, 31 Jul 2021 16:47:33 +0200 (9b3442d2f)
debian:  Do not remove users/groups in postrm - Florian Dold

Sat, 31 Jul 2021 16:08:36 +0200 (aa0f0adef)
debian: fix date - Florian Dold

Sat, 31 Jul 2021 12:58:00 +0200 (e3dc68562)
debian: bump version to 0.14.1-9 - Florian Dold

Fri, 30 Jul 2021 22:20:05 +0200 (dccf60f89)
config: handle syntax error instead of dereferencing NULL pointer - Florian Dold

Fri, 2 Apr 2021 15:46:25 +0200 (ebc70e1bc)
SETU: Implement LSD0003 - Martin Schanzenbach

Thu, 29 Jul 2021 14:26:22 -0400 (72250781a)
use modern spelling for LLONG_{MAX,MIN} - Thien-Thi Nguyen

Thu, 29 Jul 2021 17:15:32 +0200 (6212bce41)
config: restrict directives in files loaded via @inline-secret@ - Florian Dold

Thu, 29 Jul 2021 10:42:18 -0400 (12e29f004)
add dependency lib to a couple targets - Thien-Thi Nguyen

Thu, 29 Jul 2021 16:13:16 +0200 (b69774cba)
do not warn when GNUNET_DISK_file_test can't access file - Florian Dold

Thu, 29 Jul 2021 15:40:36 +0200 (73ae1f379)
config: prettier pretty printing - Florian Dold

Thu, 29 Jul 2021 15:03:19 +0200 (8af5277c8)
debian: bump version - Florian Dold

Thu, 29 Jul 2021 14:59:10 +0200 (b41bca631)
fall back to /etc for default configuration entry point - Florian Dold

Wed, 28 Jul 2021 21:23:44 +0200 (cadca90ca)
debian: bump version - Florian Dold

Wed, 28 Jul 2021 21:00:30 +0200 (3012383a1)
config: report stats for secret sections - Florian Dold

Wed, 28 Jul 2021 20:19:38 +0200 (346706b18)
config: extended diagnostics, import cycle detection, deterministic load order - Florian Dold

Wed, 28 Jul 2021 18:59:07 +0200 (1af9f6c1c)
config: sort globbed files - Florian Dold

Wed, 28 Jul 2021 17:34:44 +0200 (7a0caef7a)
diagnostics mode for gnunet-config - Florian Dold

Wed, 28 Jul 2021 15:47:39 +0200 (2cdd91d55)
fall back to /etc if user config can't be found - Florian Dold

Wed, 28 Jul 2021 15:26:55 +0200 (94ea7a1fa)
implement @inline-secret@ directive - Florian Dold

Wed, 28 Jul 2021 14:31:38 +0200 (7615d46b0)
implement config inline globbing - Florian Dold

Mon, 26 Jul 2021 20:30:28 +0200 (28dfb1220)
config: resolve realpath of parent config file, fix docs - Florian Dold

Mon, 26 Jul 2021 20:03:29 +0200 (51e5fab47)
config: free real basepath - Florian Dold

Mon, 26 Jul 2021 19:59:29 +0200 (cda72d5c4)
config: resolve realpath of base dir, not config - Florian Dold

Mon, 26 Jul 2021 19:35:56 +0200 (473939543)
debian: bump version - Florian Dold

Mon, 26 Jul 2021 19:33:38 +0200 (16fe9a808)
config: resolve symlinks for @INLINE@-ed files - Florian Dold

Mon, 26 Jul 2021 17:49:43 +0200 (302e97b01)
debian: bump version - Florian Dold

Mon, 26 Jul 2021 16:32:34 +0200 (90095d19b)
implement typed JSON packer - Christian Grothoff

Mon, 26 Jul 2021 15:45:39 +0200 (baa855368)
debian: bump version - Florian Dold

Mon, 26 Jul 2021 15:31:14 +0200 (51990dbbf)
JSON: add draft for new GNUNET_JSON_PACK API - Christian Grothoff

Sat, 24 Jul 2021 22:50:22 +0200 (d63771f25)
minimal test for new pq event functionality - Christian Grothoff

Sat, 24 Jul 2021 22:05:46 +0200 (31eae6bbe)
early draft for libgnunetpq event notification support - Christian Grothoff

Fri, 23 Jul 2021 20:10:45 +0200 (be8bca817)

src/util/gnunet-config.c
/*
     This file is part of GNUnet.
     Copyright (C) 2012-2021 GNUnet e.V.

     GNUnet is free software: you can redistribute it and/or modify it
     under the terms of the GNU Affero General Public License as published
     by the Free Software Foundation, either version 3 of the License,
     or (at your option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Affero General Public License for more details.

     You should have received a copy of the GNU Affero General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */

/**
 * @file util/gnunet-config.c
 * @brief tool to access and manipulate GNUnet configuration files
 * @author Christian Grothoff
 */
#include "platform.h"
#include "gnunet_util_lib.h"


/**
 * Backend to check if the respective plugin is
 * loadable. NULL if no check is to be performed.
 * The value is the "basename" of the plugin to load.
 */
static char *backend_check;


/**
 * If printing the value of CFLAGS has been requested.
 */
static int cflags;


/**
 * If printing the value of LIBS has been requested.
 */
static int libs;


/**
 * If printing the value of PREFIX has been requested.
 */
static int prefix;


/**
 * Print each option in a given section.
 * Main task to run to perform operations typical for
 * gnunet-config as per the configuration settings
 * given in @a cls.
 *
 * @param cls closure with the `struct GNUNET_CONFIGURATION_ConfigSettings`
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving,
 *                                                     can be NULL!)
 * @param cfg configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *cfg)
{
  struct GNUNET_CONFIGURATION_ConfigSettings *cs = cls;

  if (1 == cflags || 1 == libs || 1 == prefix)
  {
    char *prefixdir = GNUNET_OS_installation_get_path (GNUNET_OS_IPK_PREFIX);
    char *libdir = GNUNET_OS_installation_get_path (GNUNET_OS_IPK_LIBDIR);

    if (1 == cflags)
    {
      fprintf (stdout, "-I%sinclude\n", prefixdir);
    }
    if (1 == libs)
    {
      fprintf (stdout, "-L%s -lgnunetutil\n", libdir);
    }
    if (1 == prefix)
    {
      fprintf (stdout, "%s\n", prefixdir);
    }
    cs->global_ret = 0;
    GNUNET_free (prefixdir);
    GNUNET_free (libdir);
    return;
  }
  if (NULL != backend_check)
  {
    char *name;

    GNUNET_asprintf (&name,
                     "libgnunet_plugin_%s",
                     backend_check);
    cs->global_ret = (GNUNET_OK ==
                      GNUNET_PLUGIN_test (name)) ? 0 : 77;
    GNUNET_free (name);
    return;
  }
  GNUNET_CONFIGURATION_config_tool_run (cs,
                                        args,
                                        cfgfile,
                                        cfg);
}


/**
 * Program to manipulate configuration files.
 *
 * @param argc number of arguments from the command line
 * @param argv command line arguments
 * @return 0 ok, 1 on error
 */
int
main (int argc,
      char *const *argv)
{
  struct GNUNET_CONFIGURATION_ConfigSettings cs = {
    .api_version = GNUNET_UTIL_VERSION,
    .global_ret = EXIT_SUCCESS
  };
  struct GNUNET_GETOPT_CommandLineOption options[] = {
    GNUNET_GETOPT_option_exclusive (
      GNUNET_GETOPT_option_string (
        'b',
        "supported-backend",
        "BACKEND",
        gettext_noop (
          "test if the current installation supports the specified BACKEND"),
        &backend_check)),
    GNUNET_GETOPT_option_flag (
      'C',
      "cflags",
      gettext_noop (
        "Provide an appropriate value for CFLAGS to applications building on top of GNUnet"),
      &cflags),
    GNUNET_GETOPT_option_flag (
      'j',
      "libs",
      gettext_noop (
        "Provide an appropriate value for LIBS to applications building on top of GNUnet"),
      &libs),
    GNUNET_GETOPT_option_flag (
      'p',
      "prefix",
      gettext_noop (
        "Provide the path under which GNUnet was installed"),
      &prefix),
    GNUNET_CONFIGURATION_CONFIG_OPTIONS (&cs),
    GNUNET_GETOPT_OPTION_END
  };
  enum GNUNET_GenericReturnValue ret;

  if (GNUNET_OK !=
      GNUNET_STRINGS_get_utf8_args (argc, argv,
                                    &argc, &argv))
    return EXIT_FAILURE;
  ret =
    GNUNET_PROGRAM_run (argc,
                        argv,
                        "gnunet-config [OPTIONS]",
                        gettext_noop ("Manipulate GNUnet configuration files"),
                        options,
                        &run,
                        &cs);
  GNUNET_free_nz ((void *) argv);
  GNUNET_CONFIGURATION_config_settings_free (&cs);
  if (GNUNET_NO == ret)
    return 0;
  if (GNUNET_SYSERR == ret)
    return EXIT_INVALIDARGUMENT;
  return cs.global_ret;
}


/* end of gnunet-config.c */
